### PR TITLE
KAFKA-4376: Cross compile to Scala 2.12.0

### PR DESCRIPTION
**Pull request by leachbj**

### CODE REVIEW

**Comment by leachbj** [comment]
See also https://github.com/apache/kafka/pull/2088 for the 0.10.1 branch

**Comment by manub** [comment]
Hi! I'm the maintainer of a library that embeds Kafka for using with Scalatest and I'm planning to release for 2.12.

Is there any plan to merge this PR?

Thanks for the great work.

**Comment by leachbj** [comment]
@manub yeah that was pretty much my motivation here, need to get embedded Kafka updated so we can get reactive Kafka updated.

File: gradle/dependencies.gradle [hunk 1]
```diff
@@ -39,9 +39,9 @@ versions += [
   powermock: "1.6.4",
   reflections: "0.9.10",
   rocksDB: "4.11.2",
-  scalaTest: "2.2.6",
+  scalaTest: "3.0.0",
```
Comment (ijuma) [inline gradle/dependencies.gradle]: 3.0.1 has been released since the PR was created, so we can probably go straight to that.

**Review by ijuma — COMMENTED** [review]
The issue is that Scala 2.12 requires Java 8 and if we merge the change as it is, it will break Jenkins jobs that run with Java 7 as well as anyone who runs any of the `*All` commands with Java 7.

Given that, the solution is probably a bit more complicated while we still support Java 7.

One option is to only add the `2_12` variants if the build is executed with Java 8. Another option is not to add `2_12` to the `*All` commands, and use -PscalaVersion=2.12 to build for 2.12.

Thoughts?

File: gradle/dependencies.gradle (hunk 1 above)
Comment (leachbj) [reply gradle/dependencies.gradle]: Keeping the existing 2.10/2.11 builds as java7 and adding a new build for 2.12 as java 8 makes sense.
Is the jenkins job configuration part of the repo?  Or do we need to get a new job setup for jdk8?

**Comment by ijuma** [comment]
Thanks, the updated change looks good. I'll test a bit and then merge if it all goes well. We already have a JDK 8 build: https://builds.apache.org/job/kafka-trunk-jdk8. I'll update it to also build with Scala 2.12 once this PR is merged.

**Comment by leachbj** [comment]
I haven't rebased or squashed this change?  Not sure what your normal practice is - I'll leave as is unless you direct otherwise.

**Comment by ijuma** [comment]
We use a merge script that squashes all the commits into a single commit.

**Comment by ijuma** [comment]
LGTM, merged to trunk.

**Comment by manub** [comment]
Hi! Thanks for merging this. When will the 2.12 version be published on Maven Central?

**Comment by avakhrenev** [comment]
@manub according to [the plan](https://cwiki.apache.org/confluence/display/KAFKA/Time+Based+Release+Plan) it is January 2017

**Comment by leachbj** [comment]
@manub I have raised https://github.com/apache/kafka/pull/2164 to get this change onto the 0.10.1.x patch release in order to get timely access to the kafka libs on 2.12.

**Comment by manub** [comment]
Thanks @leachbj - this will unblock a bunch of downstream libraries! Keeping an eye on this.

**Pull request closed without merging**